### PR TITLE
HW2 update

### PR DESCRIPTION
**Pull request by PbX8217**

### CODE REVIEW

**Comment by linhaowei1** [comment]
Please resolve conflicts. This repo is refactored to support per-homework venv.

**Review by linhaowei1 — COMMENTED** [review]
Left some comments

File: hw2_huggingface/README.md [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+5. `AutoConfig`, `AutoTokenizer`, `AutoModelFor???`. These auto classes are available for instantiating a base model class without a specific head. Different `AutoModel` are designed for different NLP tasks, and you should choose the proper one.
+6. `datasets.map` and `AutoTokenizer` to process your dataset into tokenized IDs.
+7. `evaluate`. You need to compute `accuracy`, `micro_f1`,`macro_f1` and `weight_f1` using [this package](https://huggingface.co/docs/evaluate/index).
+8. `DataCollatorWithPadding`. Why you should use data collator before training?
+9. `Trainer` and `TrainingArguments`. This is the main training engine of 🤗transformers.
+10. `wandb`. It can track your experiments and easily interact with 🤗Huggingface. You may [sign up an account](http://wandb.ai) then read the [documents](https://docs.wandb.ai/guides/integrations/huggingface) to setup your recording. Use offline mode if you are unable to connect `wandb` server on your machine.
+
+After you finish the script,
+
+- Use `bert-base-uncased`, `facebook/bart-base` and `Qwen/Qwen1.5-0.5B` as your base model.
```
[the remark this answers is not in the document]
Comment (PbX8217) [reply hw2_huggingface/README.md:156]: I've tested that `Qwen` series is available by adding a padding token. I set pad_token=eos_token and it can converge on agnews dataset prepared in task 1. However, I didn't consider whether it would affect the calculation of attention.

[the remark this answers is not in the document]
Comment (linhaowei1) [reply hw2_huggingface/README.md:156]: ok, then it will be fine if the metrics look good

File: hw2_huggingface/README.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+uv sync --extra torch --extra metrics --extra deepspeed --prerelease=allow
+
+# for HUAWEI Ascend
+uv sync --extra torch-npu --extra metrics --extra deepspeed --prerelease=allow
+
+# set FORCE_TORCHRUN to activate deepspeed
+FORCE_TORCHRUN=1 uv run --prerelease=allow llamafactory-cli train examples/train_lora/llama3_lora_pretrain.yaml
+```
+
+If you encounter any problems about dependencies and environment setup in task 4, please turn to use conda environment and re-install LLaMA-Factory.
```
[the remark this answers is not in the document]
Comment (PbX8217) [reply hw2_huggingface/README.md:49]: I once encountered that llamafactory with deepspeed might raise some errors that the child process cannot find dependencies. It should be solved by setting two uv venv separately in hw2 root directory and llamafactory directory, I left a note here just in case.

File: hw2_huggingface/README.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+   - Set `cutoff_len: 2048`
+   - Set `lora_target: all`
+   - Set `max_samples: 3000`
+   - You can find example files in `LLama-Factory/examples/train_lora` or [here](https://ascend.github.io/docs/sources/llamafactory/quick_start.html).
+4. Write a `yaml` file to merge your lora module model with the original model.
+   - You can find example files in `LLama-Factory/examples/merge_lora`.
+5. Start the engine to fine-tune your model.
+6. Implement `eval.py` to calculate `BLEU`, `ROUGE-L` and `BERTScore-F1`, and run your evaluation on the test set.
+   - Set `facebook/bart-large` as the scoring model of `BERTScore`
+   - `bert_score` package supports batch process of multiple reference-candidate pairs, but `rouge` and `bleu` in nltk can only process one pair per call.
```
[the remark this answers is not in the document]
Comment (PbX8217) [reply hw2_huggingface/README.md:199]: I'll add some notes about adding wandb package into llamafactory venv and ask them to provide a training process recorded on wandb, and I'll test if it is feasible. Alternatively, they can submit a training log file.
